### DISCUSSION TITLE
update a Key wherever it present in given JSON

### DISCUSSION THREAD

**Discussion by SavinayaShetty** [schibsted/jslt#213] [Q&A] [answered]
Is it possible to update a key wherever it is present, irrespective of the structure of JSON?
In below example, I want to update the key "postback".
Example:
input:
```
{
  "menu": {
    "popup": {
      "menuitem": [
       {
          "action": "Open",
          "postback": "chose_open"
        },
        {
          "action": "Close",
          "postback": "chose_close"
        }
      ]
    },
   "option": {
         "action" : "OptOut",
         "postback": "not_intrested"
    }
  },
 "action": "more",
 "postback":"know_more"
}

```
output:
```
{
  "menu" : {
    "popup" : {
      "menuitem" : [ {
        "postback" : "USER1~chose_open",
        "action" : "Open"
      }, {
        "postback" : "USER1~chose_close",
        "action" : "Close"
      } ]
    },
    "option" : {
      "postback" : "USER1~not_intrested",
      "action" : "OptOut"
    }
  },
  "postback" : "USER1~know_more",
  "action" : "more"
}
```

jslt template used:
```
{
   "menu":{
      "popup":{
         "menuitem":[ for(.menu.popup.menuitem)
            {
               "postback":"USER1~"+.postback, 
               *:.
            }
         ]
      },
      "option":{
         "postback":"USER1~"+.menu.option.postback,
         *:.
      }
   },
   "postback":"USER1~"+.postback,
   *:.
}
```

Is there a more elegant way of doing this? 
In the actual case, the USER1~ will be an external parameter I'll be passing to the template, and the input JSON will be more complex.
I'm fairly new to JSLT, just wondering if any feature is available to update all in one go instead of traversing.

**Comment by larsga** [comment, marked as answer]
This is an interesting feature suggestion, but I'm afraid there is no support for anything like this at the moment.

However, you can get what you want by writing a recursive traversal function that does the update on every object in the input. You can see some useful examples here https://github.com/schibsted/jslt/blob/master/tutorial.md#function-declarations

**Reply by SavinayaShetty** [reply]
Thank you!!

**Reply by akashvekariya** [reply]
I don't think it's possible in current version at all, I came around the same problem where somehow I achieved recursive field deletion and data updation but I can't seem to find solution to key updation.

Here's some code for recursive data updation:
```
def updateByKey(jsonObj, searchTerm, updateData)
   if (is-object($jsonObj))
      {for ($jsonObj)
        .key:
        if ((is-object(.value) or is-array(.value)) and .key != $searchTerm)
           updateByKey(.value, $searchTerm, $updateData)
        else if (.key == $searchTerm)
           $updateData
        else 
           .value
      }
   else if (is-array($jsonObj))
      [for($jsonObj)
         updateByKey(., $searchTerm, $updateData)
      ]
```

Hey @larsga, can you please help me build the function correctly? 

ABOUT PROBLEM:
A recursive function which takes 3 parameters: jsonData, searchTerm, updateKeyName
What it does it find any field with the matching name as searchTerm irrespective of it's position in payload.

**Reply by larsga** [reply]
That function looks correct to me. I tested it briefly, and it worked fine. What is the problem?

**Reply by akashvekariya** [reply]
This function is to find specific key and update the data at that key location. What I want to do is, the same way I want to find the key and update the value of key itself. 

consider
`{ "age": 30 }` to `{ "myAge": 30 }`

this too, recursively. Irrespective of where this field is in JSON, it should rename every field in json at any depth

**Reply by larsga** [reply]
I see. You can do that by changing this part:

```
{for ($jsonObj)
        .key:
```

The `.key` there is an expression, so you can replace that with an `if` that returns the key name you want.

Of course, you also have to modify the rest of the function so that it doesn't replace the value.

**Reply by akashvekariya** [reply]
There exactly is a catch (or atleast I can't think of any solution to work). I want to know if I loop inside the object just like you saw, I'll need to put .key after the for loop condition so how can I check if the key matches and go back and rename or if it doesn't match with search parameter, return as it is.

Something like this I want to do, but apparently this is syntactically wrong.
```
def renameByKey(jsonObj, searchTerm, updatedKey)
   if (is-object($jsonObj))
      {for ($jsonObj)
        if ((is-object(.value) or is-array(.value)) and .key != $searchTerm)
           renameByKey(.value, $searchTerm, $updatedKey)
        else if (.key == $searchTerm)
           $updatedKey : .value
        else 
           .key : .value
      }
   else if (is-array($jsonObj))
      [for($jsonObj)
         renameByKey(., $searchTerm, $updatedKey)
      ]
```

**Reply by larsga** [reply]
Yes, the language just doesn't work that way. You're writing as if `a : b` is a value, but it's not. There are two values, separated by `:`.

It's like this: 

```
{for ($jsonObj) 
  // here .key and .value are set
<expression to produce new key> : <expression to produce new value>
}
```

So what you need is the appropriate `if` on both sides of the `:`.

However, if you *really* want, you can make a `[for ...]` that produces `[key, value]` pairs, then pass that to a `{for (...) .[0] : .[1]}`. That will allow you to write the code that way you want.

**Reply by larsga** [reply]
This should do it:

```
def renameByKey(jsonObj, searchTerm, updatedKey)
   if (is-object($jsonObj))
      {for ($jsonObj)
        if (.key != $searchTerm) .key else $updatedKey
        :
        if ((is-object(.value) or is-array(.value)) and .key != $searchTerm)
           renameByKey(.value, $searchTerm, $updatedKey)
        else 
           .value
      }
   else if (is-array($jsonObj))
      [for($jsonObj)
         renameByKey(., $searchTerm, $updatedKey)
      ]
   else
      $jsonObj
```

**Reply by akashvekariya** [reply]
Hey, thanks man! exactly what I was looking for. Works just fine!
